src/server/frontend_wayland/xdg_decoration_unstable_v1.cpp

```diff
@@ -32,6 +32,31 @@ namespace mir
 {
 namespace frontend
 {
+class ToplevelsWithDecorations
+{
+public:
+    ToplevelsWithDecorations() :
+        toplevels_with_decorations{std::make_shared<std::unordered_set<wl_resource*>>()}
+    {
+    }
+
+    /// \return true if no duplicates existed before insertion, false otherwise.
+    bool registerToplevel(wl_resource* toplevel)
+    {
+        auto [_, inserted] = toplevels_with_decorations->insert(toplevel);
+        return inserted;
+    }
+
+    /// \return true if only one element was erased, false otherwise.
+    bool unregisterToplevel(wl_resource* toplevel)
+    {
+        return toplevels_with_decorations->erase(toplevel) == 1;
+    }
+
+private:
+    std::shared_ptr<std::unordered_set<wl_resource*>> toplevels_with_decorations;
+};
```

Comment:
As you later capture `toplevels_with_decorations` by reference, nothing ever copies a `ToplevelsWithDecorations`, so there's no point to using a shared pointer here.